### PR TITLE
CACHEBOX-46 Cachebox concurrency

### PR DESCRIPTION
**Pull request by DominicWatson**
The following changes are focused on Cachebox but also touch on a couple of other uses of `java.util.concurrent.ConcurrentHashMap` in Coldbox.

The crux is that, in Lucee at least, using CFML Struct functions on instances of `java.util.concurrent.ConcurrentHashMap` are *not thread safe*, negating the purpose of using it. For instance, our application error logs show sporadic instances of `ArrayIndexOutOfBounds` and other `variable does not exist` errors in Cachebox:

```cfc
component {
// ...
    instance.pool = createObject( "java", "java.util.concurrent.ConcurrentHashMap" ).init();
// ...
    function getKeys() {
        return StructKeyArray( instance.pool ); // <<< throws ArrayIndexOutOfBounds error while pool is having keys cleared
    }
}
```

This change, then, makes use of the java methods directly defined on `java.util.concurrent.ConcurrentHashMap` to achieve what is asked of it. i.e. getting the number of keys, getting an array of keys, getting and setting values, checking key existence. It also removes the exclusive locks, defined in the cache stores, that should be redundant due to the thread safety of the `ConcurrentHashMap`. As an extra bonus, the methods to get an array of keys performed around 3x faster in my tests that `StructKeyArray()`.

We have already made these changes in Preside applications (thanks to the modular architecture of Cachebox, we could just use a separate provider that looks for our own Store) - and can see the easily reproducable thread safety errors disappear.

### CODE REVIEW

**Comment by DominicWatson** [comment]
Looks like the Travis CI build failing, but due to a build error, rather than test failure (tests run fine for me).

**Comment by bdw429s** [comment]
I'm curious if there is still a reason to use java classes at all over just a regular CFML struct.

**Comment by DominicWatson** [comment]
Ha, indeed :). I could run the tests I did to show the concurrency problems and use plain CFML structs and see if they go away. The basic error was to build a ConcurrentHashMap with loads of keys, then continually remove keys from it while adding more keys and running `StructKeyArray( myMap )`. Very quickly throws an `ArrayIndexOutOfBounds` error.

**Comment by DominicWatson** [comment]
I would have _thought_ that using the Java methods directly here would be faster though. Cuts out the layers of abstraction that Lucee adds. All should be tested though.

**Comment by bdw429s** [comment]
Well, I'm talking about just the underlying data structure and no so much the method invocation but I guess it's all tied together.  My assumption is that calling built in CFML methods on CFML data structures will be faster than calling Java methods on java classes which requires reflection.  But I'm also assuming Luis used the Java class in the first place because he felt the underlying management of the data was faster, not so much the actual method calls we make.  But I suppose maybe a test is in order to see if we're actually getting anything out of the extra complexity (and unintended bugs)

File: home/dom/code/coldbox-platform/tests/logs/commandbox.log [hunk 1]
```diff
@@ -0,0 +1,2 @@
+"Severity","Appender","Date","Time","Category","Message"
```
[the remark this answers is not in the document]
Comment (DominicWatson) [reply home/dom/code/coldbox-platform/tests/logs/commandbox.log]: doh, yes - sorry missed that. Let me delete and add ignore rule

**Comment by lmajano** [comment]
@DominicWatson all tests are failing on Adobe!

**Comment by lmajano** [comment]
@DominicWatson we have adobe incompatibilities. I can't merge this

**Comment by lmajano** [comment]
I am merging and I will modify locally

**Pull request merged**